---
title: "Data analysis"
author: "Aoyue Bi"
date: "`r Sys.Date()`"
output: 
  html_document: # html_notebook, html_document
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: true
    theme: simplex # default, united, cerulean, journal, flatly, darkly, readable, spacelab, united, cosmo, lumen, paper, sandstone, simplex, and yeti
    highlight: haddock # default, tango, pygments, kate, monochrome, espresso, zenburn, haddock, and breezedark
    # bibliography: bibliography.json
---

# Human genome

```{r setup, echo =T, message = T, warning = F}
library(tidyverse)
# library(RColorBrewer)
# library(ggpubr)
# library(egg)
## eval 代码要不要运行
## echo 代码要不要输出
## include 图要不要输出
## warning 警告要不要输出
## message 默认Bin等信息要不要输出
```

## rename file name

### step1

```{r}
### 建立哈希表
dfhm <- tibble(commonName = c("human","chimpanzee","gorilla","orangutan"),
               speciesName = c("homo_sapiens","pan_troglodytes","gorilla_gorilla","pongo_abelii"),
               abbreName = c("hg38","chim","gori","oran"))

### 数据处处理
inputDir <- c("/Users/Aoyue/project_pos/human_ILS/001_data/001_genomeInfo")

df <- dir(inputDir,full.names = T) %>% 
  map(~{read_tsv(.x) %>% 
  mutate(speciesName = str_split_fixed(basename(.x),"_sequence_report",2)[,1])
  }) %>% 
  bind_rows() %>% 
  filter(str_detect(`RefSeq seq accession`,"NC")) %>% 
  filter(!`Chromosome name` == "MT") %>%
  ### 和哈希表连接
  left_join(dfhm,by = c("speciesName"))
  
# write_tsv(df,"/Users/Aoyue/project_pos/human_ILS/001_data/002_filterfrom001/001_sequence_report.tsv")


### 下面的代码太繁琐，弃用！！！以后尽量以文件夹为主，进行文件的整合
# df_human <- read_tsv("/Users/Aoyue/project_pos/human_ILS/001_data/001_genomeInfo/homo_sapiens_sequence_report.tsv") %>% mutate(commonName = "human", speciesName = "")
# 
# df_chimpanzee <- read_tsv("/Users/Aoyue/project_pos/human_ILS/001_data/001_genomeInfo/pan_troglodytes_sequence_report.tsv")
# 
# df_gorilla <- read_tsv("/Users/Aoyue/project_pos/human_ILS/001_data/001_genomeInfo/gorilla_gorilla_sequence_report.tsv")
# 
# df_orangutan <- read_tsv("/Users/Aoyue/project_pos/human_ILS/001_data/001_genomeInfo/pongo_abelii_sequence_report.tsv")
# 
# df <- df_human %>% bind_rows(df_chimpanzee,df_gorilla,df_orangutan) %>% 
#   filter(str_detect(`RefSeq seq accession`,"NC"))
# 
# # write_tsv(df,"/Users/Aoyue/project_pos/human_ILS/001_data/002_filterfrom001/001_sequence_report.tsv")

```

### step2 rename shell

```{r}
dfcmd <- df %>% 
  mutate(CMD = str_c("mv ",commonName,".part_",`RefSeq seq accession`,".fa ", commonName,"_chr",`Chromosome name`,".fa")) %>% 
  select(CMD) %>% 
  write_tsv(.,path = "~/Documents/sh.sh",col_names = F)

```

## Preparation
- 教程学习
ucsc官网教程：
https://genomewiki.ucsc.edu/index.php/Whole_genome_alignment_howto
roast 合并序列比对文件参考网址（chenBio的博客）：
https://www.biochen.org/cn/blog/2021/%E5%A4%9A%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9/
multiple whole genome alignment 扇子和杯子：（他给出了很多的学习资源，我在这里学到了很多东西。）
https://www.jianshu.com/p/be63d73362f5

Problem running TBA for MSA:（这个网站告诉我maf文件要改为 *sing.maf,并且要使用roast，弃用tba）
https://www.seqanswers.com/forum/bioinformatics/bioinformatics-aa/26371-problem-running-tba-for-msa

kent.git官网的代码：（可以用来学习shell资源）
https://genome-source.gi.ucsc.edu/gitlist/kent.git/blob/master/src/hg/makeDb/doc/geoFor1.txt

- 注意点： 1.每条染色体建立一个文件夹。2.multiz 和 toast 命令使用的输入文件maf不能压缩，否则报错。


- create .2bit and .sizes files
```{bash eval=FALSE, include=TRUE}
### 先压缩fa为gz格式： gzip test.fa
faToTwoBit human_chr21.fa human_chr21.2bit
faToTwoBit chimpanzee_chr21.fa chimpanzee_chr21.2bit
faToTwoBit gorilla_chr21.fa gorilla_chr21.2bit
faToTwoBit orangutan_chr21.fa orangutan_chr21.2bit

faSize -detailed human_chr21.fa human_chr21.sizes
faSize -detailed chimpanzee_chr21.fa chimpanzee_chr21.sizes
faSize -detailed gorilla_chr21.fa gorilla_chr21.sizes
faSize -detailed orangutan_chr21.fa orangutan_chr21.sizes

### 测试：如果去掉./，则echo不能加引号，否则报错，为什么？
for i in ./*_chr21.fa; do echo $i; done
for i in ./*_chr21.fa; do echo $i | sed -e s/.\\///; done
### 测试：修改名字
for i in *_chr21.fa; do echo $i; `basename $i .fa`.bit; done

### 测试：输出文件的最短名字，不包含文件路径
for i in 002_refSeq_test/*chr21.fa; do echo $i; done
for i in 002_refSeq_test/*chr21.fa; do echo `basename $i`; done

### 查看文件名
for i in *chr21.fa; do echo $i; done
### 只是打印命令行,后续将命令写入脚本中，注意 >> 是追加的意思
for i in *chr21.fa; do echo "faSize -detailed $i > `echo $i | sed -e s/.fa/.sizes/`" ; done
for i in *chr21.fa; do echo "faSize -detailed $i > `echo $i | sed -e s/.fa/.sizes/`" >> sh.sh; done
### 执行命令行
for i in *chr21.fa; do faSize -detailed $i > `echo $i | sed -e s/.fa/.sizes/` ; done
### 查看上述命令执行的结果文件
for i in *_chr21.sizes; do cat $i; done

```

## script test 4 Alignments with lastz


```{bash eval=FALSE, include=TRUE}
date && echo ' lastz for human of chr21 begins\n'

lastz human_chr21.fa chimpanzee_chr21.fa --step=19 --hspthresh=2200 --inner=2000 --ydrop=3400 --gappedthresh=10000 --format=axt --ambiguous=iupac --ambiguous=n > hg38_chr21_vs_chim_chr21.axt

lastz human_chr21.fa gorilla_chr21.fa --step=19 --hspthresh=2200 --inner=2000 --ydrop=3400 --gappedthresh=10000 --format=axt --ambiguous=iupac --ambiguous=n > hg38_chr21_vs_gori_chr21.axt

lastz human_chr21.fa orangutan_chr21.fa --step=19 --hspthresh=2200 --inner=2000 --ydrop=3400 --gappedthresh=10000 --format=axt --ambiguous=iupac --ambiguous=n > hg38_chr21_vs_oran_chr21.axt

date && echo ' lastz for human of chr21 ends \n'


```

## script test 4 Chaining
```{bash eval=FALSE, include=TRUE}
### 

date && echo ' axtChain for human of chr21 begins\n'

axtChain -minScore=5000 -linearGap=loose hg38_21_vs_chim_21.axt -faT ../human_chr21.fa -faQ ../chimpanzee_chr21.fa  hg38_chr21_vs_chim_chr21.chain

axtChain -minScore=5000 -linearGap=loose hg38_21_vs_gori_21.axt -faT ../human_chr21.fa -faQ ../gorilla_chr21.fa  hg38_chr21_vs_gori_chr21.chain

axtChain -minScore=5000 -linearGap=loose hg38_21_vs_oran_21.axt -faT ../human_chr21.fa -faQ ../orangutan_chr21.fa  hg38_chr21_vs_oran_chr21.chain

date && echo ' axtChain for human of chr21 ends \n'
  
```

## script test 4 Netting
```{bash eval=FALSE, include=TRUE}
### chainMergeSort chain/*.chain > all.chain
### chainPreNet all.chain ref.size query.size all.sort.chain
### chainNet -rescore all.sort.chain ref.size query.size ref.prenet query.prenet -faT ref.fa -faQ query.fa -linearGap=loose 
### netSyntenic  ref.prenet ref.net

### NetFilterNonNested.perl -doScoreFilter -minScore1 50000 -keepSynNetsWithScore 5000 -keepInvNetsWithScore 5000 ca-di.ca.net > ca-di.ca.filter.net

### chainPreNet
date && echo ' chainPreNet for human of chr21 begins\n'

chainPreNet hg38_chr21_vs_chim_chr21.chain ../human_chr21.sizes ../chimpanzee_chr21.sizes hg38_chr21_vs_chim_chr21.sort.chain

chainPreNet hg38_chr21_vs_gori_chr21.chain ../human_chr21.sizes ../gorilla_chr21.sizes  hg38_chr21_vs_gori_chr21.sort.chain

chainPreNet hg38_chr21_vs_oran_chr21.chain ../human_chr21.sizes ../orangutan_chr21.sizes hg38_chr21_vs_oran_chr21.sort.chain

date && echo ' chainPreNet for human of chr21 ends \n'


### chainNet
chainNet hg38_chr21_vs_chim_chr21.sort.chain ../human_chr21.sizes ../chimpanzee_chr21.sizes hg38_chr21_vs_chim_chr21.prenet chim_chr21.prenet

chainNet hg38_chr21_vs_gori_chr21.sort.chain ../human_chr21.sizes ../gorilla_chr21.sizes hg38_chr21_vs_gori_chr21.prenet gori_chr21.prenet

chainNet hg38_chr21_vs_oran_chr21.sort.chain ../human_chr21.sizes ../orangutan_chr21.sizes hg38_chr21_vs_oran_chr21.prenet  oran_chr21.prenet


### netSyntenic
### 后续只用到taeget的net，因此这步只生成target的net
netSyntenic hg38_chr21_vs_chim_chr21.prenet hg38_chr21_vs_chim_chr21.syntenic.net
netSyntenic hg38_chr21_vs_gori_chr21.prenet hg38_chr21_vs_gori_chr21.syntenic.net
netSyntenic hg38_chr21_vs_oran_chr21.prenet hg38_chr21_vs_oran_chr21.syntenic.net

### netToAxt ref.net 
netToAxt hg38_chr21_vs_chim_chr21.syntenic.net hg38_chr21_vs_chim_chr21.sort.chain ../human_chr21.2bit ../chimpanzee_chr21.2bit  hg38_chr21_vs_chim_chr21.net.axt 

netToAxt hg38_chr21_vs_gori_chr21.syntenic.net hg38_chr21_vs_gori_chr21.sort.chain ../human_chr21.2bit ../gorilla_chr21.2bit hg38_chr21_vs_gori_chr21.net.axt

netToAxt hg38_chr21_vs_oran_chr21.syntenic.net hg38_chr21_vs_oran_chr21.sort.chain ../human_chr21.2bit ../orangutan_chr21.2bit hg38_chr21_vs_oran_chr21.net.axt


### axtSort
axtSort hg38_chr21_vs_chim_chr21.net.axt hg38_chr21_vs_chim_chr21.sort.net.axt

axtSort hg38_chr21_vs_gori_chr21.net.axt hg38_chr21_vs_gori_chr21.sort.net.axt

axtSort hg38_chr21_vs_oran_chr21.net.axt hg38_chr21_vs_oran_chr21.sort.net.axt


date && echo ' axtChain for human of chr21 ends \n'

```

## script test 4 Maffing 
- !!!!!!注意使用 multiz 命令的时候，maf文件不能压缩！！！
```{bash eval=FALSE, include=TRUE}
### axtToMaf
axtToMaf -tPrefix=hg38. -qPrefix=chim. hg38_chr21_vs_chim_chr21.sort.net.axt ../human_chr21.sizes ../chimpanzee_chr21.sizes hg38_chr21_vs_chim_chr21.maf

axtToMaf -tPrefix=hg38. -qPrefix=gori. hg38_chr21_vs_gori_chr21.sort.net.axt ../human_chr21.sizes ../gorilla_chr21.sizes hg38_chr21_vs_gori_chr21.maf

axtToMaf -tPrefix=hg38. -qPrefix=oran. hg38_chr21_vs_oran_chr21.sort.net.axt ../human_chr21.sizes ../orangutan_chr21.sizes hg38_chr21_vs_oran_chr21.maf

### !!!!注意压缩文件

### multiz
multiz hg38_chr21_vs_chim_chr21.maf.gz hg38_chr21_vs_gori_chr21.maf.gz hg38_chr21_vs_oran_chr21.maf.gz > all.maf

### 创建软连接，重新运行multiz
ln -s hg38_chr21_vs_chim_chr21.maf hg38.chim.sing.maf
ln -s hg38_chr21_vs_gori_chr21.maf hg38.gori.sing.maf
ln -s hg38_chr21_vs_oran_chr21.maf hg38.oran.sing.maf

### 两两合并，多做几次
multiz hg38.chim.maf.gz hg38.gori.maf.gz 1 > all.maf

multiz hg38.chim.maf hg38.gori.maf 1 > all.maf

multiz hg38_chr21_vs_chim_chr21.maf hg38_chr21_vs_gori_chr21.maf 1 >all.maf
multiz all.maf hg38_chr21_vs_oran_chr21.maf 1 > final.maf

### tba 方法合并maf文件,舍弃，原因很复杂，包括：1.文件命名问题2.fa文件命名问题
tba “newick_tree” *.maf tba_out.maf
tba "(((hg38.NC_000021.9 chim.NC_072419.1) gori.NC_073245.1) oran.NC_072006.1)" *.maf all.maf
tba "(((hg38. chim.) gori.) oran.)" *.maf all.maf
tba "(((hg38. chim.) gori.) oran.)" hg38.chim.maf hg38.gori.maf  hg38.oran.maf all.maf
tba "(((hg38 chim) gori) oran)" hg38.chim.sing.maf hg38.gori.sing.maf  hg38.oran.sing.maf all.maf
tba "(((hg38 chim) gori) oran)" *.*.maf tba.maf

### roast 方法成功！适合多序列的合并。
roast - T=`pwd` E=hg38 "(((hg38 chim) gori) oran)" *.*.maf tba.maf > roast.sh
bash roast.sh


### 代码解析
### 列出所有两两比对的maf文件，定义合并后的总的maf文件
maf_array=($( ls -d ./analyses/last/net_axt/net_maf/*1to1.maf ))
combined_maf=./analyses/last/net_axt/net_maf/combined.maf

### 
cat ${maf_array[@]:0:1} | sed -n '/##maf version=1 scoring=blastz/,$p' > \
${maf_array[@]:0:1}_tmp
cat ${maf_array[@]:1:1} | sed -n '/##maf version=1 scoring=blastz/,$p' > \
${maf_array[@]:1:1}_tmp

multiz ${maf_array[@]:0:1}_tmp ${maf_array[@]:1:1}_tmp 1 > $combined_maf

for maf in ${maf_array[@]:2};
do
  echo "processing " $maf
 cat $maf | sed -n '/##maf version=1 scoring=blastz/,$p' > \
 "$maf"_tmp
 multiz $combined_maf "$maf"_tmp 1 > "$combined_maf"_tmp
 mv "$combined_maf"_tmp $combined_maf
done

# and filter mafs so all blocks have Zea mays and are at least 20 bp long
mafFilter -minCol=20 -needComp="$ref_name" $combined_maf > "$combined_maf".filtered


### 自己修改脚本进行测试
maf_array=($( ls -d ./*.maf.gz ))
combined_maf=./combined.maf

cat ${maf_array[@]:0:1} | sed -n '/##maf version=1 scoring=blastz/,$p' > \
${maf_array[@]:0:1}_tmp
cat ${maf_array[@]:1:1} | sed -n '/##maf version=1 scoring=blastz/,$p' > \
${maf_array[@]:1:1}_tmp

multiz ${maf_array[@]:0:1}_tmp ${maf_array[@]:1:1}_tmp 1 > $combined_maf

for maf in ${maf_array[@]:2};
do
  echo "processing " $maf
 cat $maf | sed -n '/##maf version=1 scoring=blastz/,$p' > \
 "$maf"_tmp
 multiz $combined_maf "$maf"_tmp 1 > "$combined_maf"_tmp
 mv "$combined_maf"_tmp $combined_maf
done

# and filter mafs so all blocks have Zea mays and are at least 20 bp long
mafFilter -minCol=20 -needComp="$ref_name" $combined_maf > "$combined_maf".filtered
```

## script test 4 CoalHMM
```{bash eval=FALSE, include=TRUE}

### multiz M=1 
```


## from long zhou
```{bash eval=FALSE, include=TRUE}
for i in `seq 1 22` X Y; do mkdir chr$i; echo "/slurm/users/zhouyang/software/kentUtils/faToTwoBit /slurm/users/zhouyang/project/human_T2T/05.assembly_comparison/00.data/MF2_combine.v0.6.split/chr$i.fasta `pwd`/chr$i/target.2bit && /slurm/users/zhouyang/software/kentUtils/faToTwoBit /slurm/users/zhouyang/project/human_T2T/05.assembly_comparison/00.data/CHM13.split/chr$i.fasta `pwd`/chr$i/query.2bit && /slurm/users/zhouyang/software/kentUtils/faSize /slurm/users/zhouyang/project/human_T2T/05.assembly_comparison/00.data/MF2_combine.v0.6.split/chr$i.fasta -detailed > `pwd`/chr$i/target.sizes && /slurm/users/zhouyang/software/kentUtils/faSize /slurm/users/zhouyang/project/human_T2T/05.assembly_comparison/00.data/CHM13.split/chr$i.fasta -detailed > `pwd`/chr$i/query.sizes && lastz /slurm/users/zhouyang/project/human_T2T/05.assembly_comparison/00.data/MF2_combine.v0.6.split/chr$i.fasta /slurm/users/zhouyang/project/human_T2T/05.assembly_comparison/00.data/CHM13.split/chr$i.fasta --hspthresh=36400 --format=axt > `pwd`/chr$i/chr$i.axt && /slurm/users/zhouyang/software/kentUtils/axtChain -minScore=5000 -linearGap=/slurm/users/bixupeng/local/pipeline/lastz/medium `pwd`/chr$i/chr$i.axt `pwd`/chr$i/target.2bit `pwd`/chr$i/query.2bit `pwd`/chr$i/chr$i.chain"; done > lastz.shell
python3 ~/bin/sbatch-submit.py -c 20 --mem 50g -l 1 lastz.shell

~/software/kentUtils/chainMergeSort */*.chain > all.chain
~/software/kentUtils/chainPreNet all.chain ~/project/human_T2T/05.assembly_comparison/00.data/MF2_combine.v0.6.sizes ~/project/human_T2T/05.assembly_comparison/00.data/chm13v2.0.sizes all_sort.chain
~/software/kentUtils/chainNet all_sort.chain ~/project/human_T2T/05.assembly_comparison/00.data/MF2_combine.v0.6.sizes ~/project/human_T2T/05.assembly_comparison/00.data/chm13v2.0.sizes temp query.net
~/software/kentUtils/netSyntenic temp target.net
~/software/kentUtils/netToAxt target.net all_sort.chain ~/project/human_T2T/05.assembly_comparison/00.data/MF2_combine.v0.6.2bit ~/project/human_T2T/05.assembly_comparison/00.data/chm
```


## from fuqiang lin
```{bash eval=FALSE, include=TRUE}
#得到以法老为坐标参考的maf文件
#预备数据
faToTwoBit mpha.fa mpha.2bit  
faToTwoBit dmel.fa dmel.2bit  
faSize mpha.fa -detailed > mpha.sizes  
faSize dmel.fa -detailed > dmel.sizes

#lastz本身不支持并行，所以将法老基因组按照染色体切分，手动并行。
mkdir fa  
faSplit byName mpha.fa ./fa
mkdir axt  
for i in ./fa/*.fa; do prefix=$(basename $i .fa); lastz $i dmel.fa O=400 E=30 K=3000 L=3000 H=2200 T=1 --format=axt --ambiguous=n --ambiguous=iupac > axt/${prefix}.axt; done
#lastz fa/chr1.fa Cyprinus_carpio.fa O=400 E=30 K=3000 L=3000 H=2200 T=1 --format=axt --ambiguous=n --ambiguous=iupac > axt/chr1.axt

#chaining
mkdir chain  
for i in axt/*.axt; do prefix=$(basename $i .axt); axtChain $i mpha.2bit dmel.2bit chain/${prefix}.chain -minScore=3000 -linearGap=medium
#axtChain axt/chr1.axt Danio_rerio.2bit Cyprinus_carpio.2bit chain/chr1.chain -minScore=3000 -linearGap=medium
chainMergeSort chain/*.chain > 01.1.all.chain  
chainPreNet 01.1.all.chain mpha.sizes dmel.sizes 01.2.all.pre.chain

#netting
chainNet 01.2.all.pre.chain -minSpace=1 mpha.sizes dmel.sizes 02.1.mpha.net 02.2.dmel.net 
netSyntenic 02.1.mpha.net 03.mpha.noClass.net

#maffing
netToAxt 03.mpha.noClass.net 01.2.all.pre.chain mpha.2bit dmel.2bit 04.1.out.axt
axtSort 04.1.out.axt 04.2.mpha.dmel.axt  
axtToMaf 04.2.mpha.dmel.axt mpha.sizes dmel.sizes 05.mpha.dmel.maf -tPrefix=mpha. -qPrefix=dmel.


nohup paralleltask -t slurm -l 1 -m 10 -M 3G -p 10 --job_prefix pl_task task.sh &

良渚我装在这了：/share/home/zhanglab/user/linfuqiang/software/miniconda3/bin/paralleltask

```

